src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneJobAction.java

```diff
@@ -0,0 +1,68 @@
+package org.jenkinsci.plugins.pipeline.milestone;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.model.InvisibleAction;
+import hudson.model.Job;
+import hudson.model.Run;
+import java.io.Serial;
+import java.util.Map;
+import java.util.TreeMap;
+import net.jcip.annotations.GuardedBy;
+
+/**
+ * Stores passed milestones for currently running builds.
+ */
+public class MilestoneJobAction extends InvisibleAction {
+    /**
+     * Do not persist, live data is rebuilt by {@link MilestoneStepExecution.FlowExecutionListenerImpl}.
```

Comment:
So…why even define as an `Action` and add to the `Job`? Is there some advantage beyond simply keeping a `transient WeakHashMap<Job, Map<Integer, Integer>>>` as a field in some `Extension`?

Reply:
My mind was stuck based on my initial thought which was to persist at job level. But since we don't persist at all anymore, indeed this can be simplified.